### PR TITLE
Iaco 569 open registration webview

### PR DESCRIPTION
**Pull request by tomerlevi444**

### CODE REVIEW

File: klarna-on-demand/src/main/java/com/klarna/ondemand/WebViewActivity.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        ActionBar actionBar = getActionBar();
+        actionBar.setDisplayHomeAsUpEnabled(true);
+        actionBar.setHomeButtonEnabled(true);
+    }
+
+
+    private void showDismissAlert() {
+        final AlertDialog.Builder builder = new AlertDialog.Builder(this);
+        builder.setTitle(this.getTitle());
+        builder.setMessage("Are you sure you want to exit?");
```
Comment (tomerlevi444) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/WebViewActivity.java]: @tomeralon, I need your specification for the English and Swedish message and title (if there should be a title).

File: example/src/main/java/com/klarna/example/MainActivity.java [hunk 2]
```diff
@@ -1,41 +1,41 @@
...
 
 import android.app.Activity;
+import android.content.Intent;
 import android.os.Bundle;
-import android.view.Menu;
-import android.view.MenuItem;
+import android.view.View;
 import com.klarna.ondemand.Context;
+import com.klarna.ondemand.RegistrationActivity;
+import com.klarna.ondemand.WebViewActivity;
```
Comment (grozen) [inline example/src/main/java/com/klarna/example/MainActivity.java]: Do we really need to explicitly import this?

File: example/src/main/java/com/klarna/example/MainActivity.java [hunk 3]
```diff
@@ -1,41 +1,41 @@
...
 
 
 public class MainActivity extends Activity {
 
+    public final int REQUEST_CODE = 1;
```
Comment (grozen) [inline example/src/main/java/com/klarna/example/MainActivity.java]: This is really a bit too generic. Tie the name to the registration process, perhaps?

File: klarna-on-demand/build.gradle [hunk 4]
```diff
@@ -1,6 +1,18 @@
...
 android {
+    packagingOptions {
+        exclude 'DEPENDENCIES.txt'
+        exclude 'META-INF/LICENSE.txt'
+        exclude 'META-INF/INDEX.LIST'
+        exclude 'NOTICE.txt'
+        exclude 'META-INF/NOTICE'
+        exclude 'META-INF/LICENSE'
+        exclude 'META-INF/DEPENDENCIES'
+        exclude 'notice.txt'
```
Comment (grozen) [inline klarna-on-demand/build.gradle]: Why is this file excluded twice? (Assuming we did not mean to be case sensitive here).
Do we even have this file? What does it contain? Why don't we exclude the readme?

File: klarna-on-demand/src/androidTest/java/com/klarna/ondemand/ContextTest.java [hunk 5]
```diff
@@ -26,4 +25,4 @@ public void testGetApiKeyShouldThrowExceptionWhenThereIsNoApiKey() {
             assertEquals(e.getMessage(), "You must set the API key first.");
```
Comment (grozen) [inline klarna-on-demand/src/androidTest/java/com/klarna/ondemand/ContextTest.java]: Would you be so kind as to change the text on line 22 to something less confusing such as `"Exception not thrown"`.

File: klarna-on-demand/src/androidTest/java/com/klarna/ondemand/UrlHelperTest.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    public void testRegistrationUrl_ShouldReturnPlaygroundUrl_WhenTokenIsForPlayground() {
+        assertTrue(UrlHelper.registrationUrl().startsWith("https://inapp.playground.klarna.com/registration/new"));
+    }
+
+    public void testRegistrationUrl_ShouldReturnProductionUrl_WhenTokenIsForProduction() {
+        Context.setApiKey("skadoo");
+        assertTrue(UrlHelper.registrationUrl().startsWith("https://inapp.klarna.com/registration/new"));
+    }
+
+    public void testRegistrationUrl_ShouldReturnUrlWithSwedishLocale_WhenLocaleIsSwedish() {
```
Comment (grozen) [inline klarna-on-demand/src/androidTest/java/com/klarna/ondemand/UrlHelperTest.java]: This test does not rule out the fact that we may be adding "locale=sv" regardless of the locale. Won't you add an additional test?

File: klarna-on-demand/src/androidTest/java/com/klarna/ondemand/UrlHelperTest.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        assertTrue(UrlHelper.preferencesUrl(token).startsWith(expectedPrefix));
+    }
+
+    public void testPreferencesUrlWithToken_ShouldReturnProductionUrl_WhenTokenIsForProduction() {
+        Context.setApiKey("skadoo");
+        String expectedPrefix = String.format("%s%s%s", "https://inapp.klarna.com/users/", token, "/preferences");
+        assertTrue(UrlHelper.preferencesUrl(token).startsWith(expectedPrefix));
+    }
+
+    public void testPreferencesUrlWithToken_ShouldReturnUrlWithSwedishLocale_WhenLocaleIsSwedish() {
```
Comment (grozen) [inline klarna-on-demand/src/androidTest/java/com/klarna/ondemand/UrlHelperTest.java]: The comment from line 24 applies here as well.

File: klarna-on-demand/src/main/java/com/klarna/ondemand/Context.java [hunk 8]
```diff
@@ -1,13 +1,24 @@
 package com.klarna.ondemand;
 
+/**
+ * Klarna On Demand context
```
Comment (grozen) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/Context.java]: Maybe replace this with more flavorful text (also don't capitalize "on")?
`Manages the application wide context for Klarna on Demand`?

(I know you copied this from the ios project. That will require changing as well :stuck_out_tongue_winking_eye:)

File: klarna-on-demand/src/main/java/com/klarna/ondemand/UrlHelper.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    static String registrationUrl() {
+        return String.format("%s/registration/new?api_key=%s&locale=%s", UrlHelper.baseUrl(), Context.getApiKey(), defaultLocale());
+    }
+
+    static String preferencesUrl(String token) {
+        return String.format("%s/users/%s/preferences?api_key=%s&locale=%s", UrlHelper.baseUrl(), token, Context.getApiKey(), defaultLocale());
+    }
+
+    static String defaultLocale() {
+        return Locale.getDefault().getLanguage();
```
Comment (grozen) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/UrlHelper.java]: Does this behave "correctly", on par with what happens in iOS?

File: klarna-on-demand/src/main/java/com/klarna/ondemand/UrlHelper.java [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+    static String baseUrl() {
+        if(Context.getApiKey().startsWith("test_")) {
+            return KlarnaPlaygroundUrl;
+        }
+        return KlarnaProductionUrl;
+    }
+
+    static String registrationUrl() {
+        return String.format("%s/registration/new?api_key=%s&locale=%s", UrlHelper.baseUrl(), Context.getApiKey(), defaultLocale());
```
Comment (grozen) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/UrlHelper.java]: While do you feel the need to qualify the call to `baseUrl` but not the one to `defaultLocale`?

File: klarna-on-demand/src/main/java/com/klarna/ondemand/WebViewActivity.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        setContentView(R.layout.activity_webview);
+
+        addSpinner();
+
+        initializeActionBar();
+
+        initializeWebView();
```
Comment (grozen) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/WebViewActivity.java]: Have this return the webview and avoid the extra lookup a line below this?

File: klarna-on-demand/src/main/java/com/klarna/ondemand/WebViewActivity.java (hunk 1 above)
Comment (grozen) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/WebViewActivity.java]: Do we have a locale ticket referencing this?

**Comment by grozen** [comment]
Smashing job overall!

File: klarna-on-demand/src/main/java/com/klarna/ondemand/WebViewActivity.java (hunk 1 above)
Comment (tomerlevi444) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/WebViewActivity.java]: There is locale support.

File: example/src/main/java/com/klarna/example/MainActivity.java (hunk 3 above)
Comment (tomerlevi444) [inline example/src/main/java/com/klarna/example/MainActivity.java]: changed to: KLARNA_REGISTRATION_REQUEST_CODE

File: klarna-on-demand/build.gradle (hunk 4 above)
Comment (tomerlevi444) [inline klarna-on-demand/build.gradle]: To be honest, this is a common make-it-work solution that people tend to use with Gradle.

File: klarna-on-demand/src/androidTest/java/com/klarna/ondemand/ContextTest.java (hunk 5 above)
Comment (tomerlevi444) [inline klarna-on-demand/src/androidTest/java/com/klarna/ondemand/ContextTest.java]: Done

File: klarna-on-demand/src/main/java/com/klarna/ondemand/Context.java (hunk 8 above)
Comment (tomerlevi444) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/Context.java]: WDYT about `Manages the application-wide context for Klarna on Demand payments.` ?

File: klarna-on-demand/src/main/java/com/klarna/ondemand/UrlHelper.java (hunk 9 above)
Comment (tomerlevi444) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/UrlHelper.java]: Apparently the locale support for Android has a much simpler infrastructure - there is only the locale defined in the device.

File: klarna-on-demand/src/main/java/com/klarna/ondemand/UrlHelper.java (hunk 10 above)
Comment (tomerlevi444) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/UrlHelper.java]: Changed `UrlHelper.baseUrl()` to `baseUrl()`.

File: klarna-on-demand/src/main/java/com/klarna/ondemand/WebViewActivity.java (hunk 11 above)
Comment (tomerlevi444) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/WebViewActivity.java]: Done

File: klarna-on-demand/src/main/java/com/klarna/ondemand/Context.java (hunk 8 above)
Comment (grozen) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/Context.java]: Sounds good to me.

File: klarna-on-demand/src/main/java/com/klarna/ondemand/WebViewActivity.java (hunk 1 above)
Comment (tomeralon) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/WebViewActivity.java]: I might be able to get the English and Swedish only next week.

File: klarna-on-demand/build.gradle (hunk 4 above)
Comment (yuval-netanel) [inline klarna-on-demand/build.gradle]: I think this is more related to the jar making task.

File: klarna-on-demand/src/androidTest/java/com/klarna/ondemand/UrlHelperTest.java [hunk 12]
```diff
@@ -0,0 +1,49 @@
+package com.klarna.ondemand;
+
+import junit.framework.TestCase;
+import java.util.Locale;
+
+public class UrlHelperTest extends TestCase {
+
+    String token = "my_token";
+
+    public void setUp() throws java.lang.Exception {
```
Comment (yuval-netanel) [inline klarna-on-demand/src/androidTest/java/com/klarna/ondemand/UrlHelperTest.java]: @Override

File: klarna-on-demand/src/androidTest/java/com/klarna/ondemand/ContextTest.java [hunk 13]
```diff
@@ -14,16 +12,17 @@ protected void setUp() throws java.lang.Exception {
     public void testSetApiKey(){
```
Comment (yuval-netanel) [inline klarna-on-demand/src/androidTest/java/com/klarna/ondemand/ContextTest.java]: @Override

File: klarna-on-demand/src/androidTest/java/com/klarna/ondemand/UrlHelperTest.java [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    //region .registrationUrl
+    public void testRegistrationUrl_ShouldReturnPlaygroundUrl_WhenTokenIsForPlayground() {
+        assertTrue(UrlHelper.registrationUrl().startsWith("https://inapp.playground.klarna.com/registration/new"));
+    }
+
+    public void testRegistrationUrl_ShouldReturnProductionUrl_WhenTokenIsForProduction() {
+        Context.setApiKey("skadoo");
+        assertTrue(UrlHelper.registrationUrl().startsWith("https://inapp.klarna.com/registration/new"));
```
Comment (yuval-netanel) [inline klarna-on-demand/src/androidTest/java/com/klarna/ondemand/UrlHelperTest.java]: We should verify the api_key parameter in both tests

File: klarna-on-demand/src/androidTest/java/com/klarna/ondemand/UrlHelperTest.java [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    public void testRegistrationUrl_ShouldReturnUrlWithSwedishLocale_WhenLocaleIsSwedish() {
+        Locale.setDefault(new Locale("sv"));
```
Comment (yuval-netanel) [inline klarna-on-demand/src/androidTest/java/com/klarna/ondemand/UrlHelperTest.java]: This has effect on other test ...
- We could add after which fix the default locale
- Mock the call
- Remember to fix the locale english before other test

File: klarna-on-demand/src/androidTest/java/com/klarna/ondemand/UrlHelperTest.java [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+
+        assertTrue(UrlHelper.registrationUrl().contains("locale=sv"));
+    }
+    //endregion
+
+    //region .PreferencesUrlWithToken
+    public void testPreferencesUrlWithToken_ShouldReturnPlaygroundUrl_WhenTokenIsForPlayground() {
+        String expectedPrefix = String.format("%s%s%s", "https://inapp.playground.klarna.com/users/", token, "/preferences");
```
Comment (yuval-netanel) [inline klarna-on-demand/src/androidTest/java/com/klarna/ondemand/UrlHelperTest.java]: I prefer (the format is overkill):
 String expectedPrefix = "https://inapp.playground.klarna.com/users/" + token + "/preferences";

File: klarna-on-demand/src/androidTest/java/com/klarna/ondemand/UrlHelperTest.java [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+        assertTrue(UrlHelper.preferencesUrl(token).startsWith(expectedPrefix));
```
Comment (yuval-netanel) [inline klarna-on-demand/src/androidTest/java/com/klarna/ondemand/UrlHelperTest.java]: We should verify the api_key parameter in both tests

File: klarna-on-demand/src/main/java/com/klarna/ondemand/Context.java [hunk 18]
```diff
@@ -1,13 +1,24 @@
...
     public static void setApiKey(String apiKey){
         Context.apiKey = apiKey;
     }
 
-    protected static String getApiKey(){
+    /**
+     * Returns the API key set using setApiKey:.
+     * @return Merchant's public API key for this application.
+     */
+    static String getApiKey(){
```
Comment (yuval-netanel) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/Context.java]: Why not protected? the user doesn't need this method

File: klarna-on-demand/src/main/java/com/klarna/ondemand/WebViewActivity.java (hunk 1 above)
Comment (yuval-netanel) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/WebViewActivity.java]: Anyway move the translation to string.xml

**Comment by yuval-netanel** [comment]
Approved!

File: klarna-on-demand/src/main/java/com/klarna/ondemand/WebViewActivity.java (hunk 1 above)
Comment (tomerlevi444) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/WebViewActivity.java]: @yuval-netanel Done.

File: klarna-on-demand/src/androidTest/java/com/klarna/ondemand/UrlHelperTest.java (hunk 12 above)
Comment (tomerlevi444) [inline klarna-on-demand/src/androidTest/java/com/klarna/ondemand/UrlHelperTest.java]: Done.

File: klarna-on-demand/src/androidTest/java/com/klarna/ondemand/ContextTest.java (hunk 13 above)
Comment (tomerlevi444) [inline klarna-on-demand/src/androidTest/java/com/klarna/ondemand/ContextTest.java]: Done.

File: klarna-on-demand/src/androidTest/java/com/klarna/ondemand/UrlHelperTest.java (hunk 15 above)
Comment (tomerlevi444) [inline klarna-on-demand/src/androidTest/java/com/klarna/ondemand/UrlHelperTest.java]: Added Locale.setDefault(Locale.ENGLISH) at the end of the test.

File: klarna-on-demand/src/androidTest/java/com/klarna/ondemand/UrlHelperTest.java (hunk 16 above)
Comment (tomerlevi444) [inline klarna-on-demand/src/androidTest/java/com/klarna/ondemand/UrlHelperTest.java]: Done.

File: klarna-on-demand/src/main/java/com/klarna/ondemand/Context.java (hunk 18 above)
Comment (tomerlevi444) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/Context.java]: Correct!
I changed it (when I was evaluating Powermock) and forgot to restore.

**Pull request merged**